Stats Widget: Open Stats view when tapping the widget after the app was quit

staskus opened the pull request:
## Issue

Closes #18452 

## Description

### Issue

When Stats widget is configured and then the app is closed, tapping on the widget does not open Stats view. Tapping on the widget opens Stats view only after we first open Stats view manually in the app.

### Additional issue found while testing

[When tapping Stats widget, navigation doesn't work if Initial Screen is set to "Home"](https://github.com/wordpress-mobile/WordPress-iOS/pull/18837#issuecomment-1157561284)

### Rootcause

1. The issue happens because `showStats(..)` method in `MySitesCoordinator` exits early if `lastSelectedStatsPeriodTypeKey` doesn't have value. 
2. `lastSelectedStatsPeriodTypeKey` depends on `SiteStatsInformation.sharedInstance.siteID`. `SiteStatsInformation`values are only set when `StatsViewController` is initialized. That's why  `SiteStatsInformation.sharedInstance.siteID` is nil after quiting and reopening the app.
3. The bug is a regression from [Empty Stats: Stay on insights tab if Grow Audience card is showing](https://github.com/wordpress-mobile/WordPress-iOS/pull/17332)

### Solution

1. Make callers of `lastSelectedStatsPeriodTypeKey` pass `postID` to make it usable in context where `SiteStatsInformation` is not initialized yet.
2.  Use `blog.dotComID` in `MySitesCoordinator` when getting `lastSelectedStatsPeriodTypeKey`.

### Other Considerations

- Use `NavigationAction.defaultBlog()?` or a similar default `Blog` instance for returning a `lastSelectedStatsPeriodTypeKey`. I chose to pass siteID since we already had a `Blog` passed to a `showStats(..)` method. 
- Don't use `guard let` and don't return if `lastSelectedStatsPeriodTypeKey` doesn't exist. If we do this, stats page would open but the correct time tab wouldn't be chosen. 
- Preload `SiteStatsInformation` when site is selected. That would be a more significant change and since this is a regression I made a decision to return the code to a previous state.

## Testing instructions

1. Make sure a site is selected in WordPress app
2. Open Stats view
3. Put the app in the background
4. Add Today, This Week and All Time Widget
5. Confirm that tapping on widgets open corresponding Today, This Week and Insights Stat views
6. Force quit the app in the app switcher
7. Tap on Today widget
8. Confirm that the Stats view is opened with the corresponding Today view
9. Repeat steps 6-8 for This Week and All Time Widgets. 

## Regression Notes

1. Potential unintended areas of impact

- Breaking functionality of https://github.com/wordpress-mobile/WordPress-iOS/pull/17332
- The other possibility of some unintended consequences if there're valid cases where `blog.dotComID` **should** be different from `SiteStatsInformation.sharedInstance.siteID` but I couldn't find the reason why it should be.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- Perfomed tests described in https://github.com/wordpress-mobile/WordPress-iOS/pull/17332
- Didn't rely on any automated tests

3. What automated tests I added (or what prevented me from doing so)

I considered adding `MySitesCoordinatorTests` since there're tests for other types of coordinators but since this is my first PR I want to make and early PR and consult before doing so. 

## Images and Gif

![issue-18452](https://user-images.githubusercontent.com/4062343/172412049-8670db0d-a5d3-490a-b687-64c5ae4fd41e.gif)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Remove lastSelectedStatsPeriodTypeKey dependency on SiteStatsInformation

SiteStatsInformation is only loaded after StatsViewController is initialized.

lastSelectedStatsPeriodTypeKey depended on SiteStatsInformation
and didn't allow to use it correctly in contexts where StatsViewController hasn't been loaded yet.

guarani commented:
👋 Hi @momo-ozawa, this PR touches code you worked on last year (as described above). Would you be available to review this?

staskus commented:
@momo-ozawa 

Unfortunately, while working on the other bug I noticed that this fix is no longer working. After playing around and debugging I found that after disabling "My site dashboard" feature flag it starts working again so it looks to be related. 

I suppose I need to make additional navigational changes to support new setup. I saw you worked on these dashboard changes as well, do you have a suggestion how to better make these changes? Thanks.

![Simulator Screen Recording - iPhone 13 mini - 2022-06-14 at 15 45 01](https://user-images.githubusercontent.com/4062343/173581018-a277e68c-24ad-4df8-a109-2b25532dfa6c.gif)

momo-ozawa reviewed (commented):
Hey @staskus (and @guarani) 👋 

Oops -- didn't see your comment above when I wrote my original comment.

So what's happening in the gif is that the initial screen is set to Home (aka the dashboard). If the initial screen is set to Menu, your fix seems to be working. To change the inital screen (i.e. the default selection on the segmented control), you can tap on the profile icon on My Site > App Settings > Initial Screen. 

I tested on WPiOS 19.9.1 (current release) and I can reproduce this issue. This means that not being able to open the correct stats view when the initial screen is set to Home is _not_ a regression caused by this PR. @guarani I wonder if fixing this behavior for dashboard should be tackled in a separate PR.

@staskus In the meantime, I'll re-review your code and also think of suggestions on tackling the dashboard problem.

momo-ozawa reviewed (commented):
@staskus The fix works as described when the initial screen is set to Menu as mentioned in my comment above. 

> The other possibility of some unintended consequences if there're valid cases where blog.dotComID should be different from SiteStatsInformation.sharedInstance.siteID but I couldn't find the reason why it should be.

I think this should be fine. `blog.dotComID` should be the same as `SiteStatsInformation.sharedInstance.siteID`.

> Tackling the dashboard problem

Currently `showDetailView(for:)` in `MySiteViewController` isn't set up to handle showing the stats view from the dashboard. 

https://github.com/wordpress-mobile/WordPress-iOS/blob/0bf27fadfaa3882205dfa947c4c8dfb6ae0009aa/WordPress/Classes/ViewRelated/Blog/My%20Site/MySiteViewController.swift#L972-L978

We'd need to refactor this method to route to the stats screen via `BlogDashboardViewController`. The child vc for `MySiteViewController` depends on which tab is selected on the segmented control:

- Home -> `BlogDashboardViewController`
- Menu -> `BlogDetailsViewController`

Hope that helps! I'll defer to @guarani about whether it's better to tackle the dashboard issue on this PR or not. Also, thank you for the thorough investigation notes! 😄

staskus commented:
@momo-ozawa Thank you for your comments and suggestions!

guarani commented:
Hi @momo-ozawa, thank you for helping out! ❤️ 

---

Hi @staskus 👋 , it sounds like the bug that this PR addresses, https://github.com/wordpress-mobile/WordPress-iOS/issues/18452, affects both users who have the Home tab as default, as well as those who have the Menu tab as default. Furthermore, a separate navigation issue prevents this PR from fixing the case where the Home tab is default. Is that an accurate understanding?

Please feel free to make a call on whether to include the "Home tab as default" scenario under https://github.com/wordpress-mobile/WordPress-iOS/issues/18452, or create a separate issue. Either way, would you be able to make a fix for it as well?

staskus commented:
@guarani, yes, you understand it correctly. There's a separate navigation issue not entirely related to the reported issue, that prevents deep-link navigation to work from the Home tab.

I'll fix the navigation issue in this PR so the fix would fully cover both scenarios.

guarani commented:
Thank you @staskus!

staskus committed:
Use single entry point for showing blog details

- Widget router uses mySiteViewController.showDetailsView
- WP3DTouchShortcutHandler uses mySiteViewController.showBlogDetailsSubsection

These methods have exactly same functionality. Use only one of them  to avoid duplication and errors that come from it.

staskus committed:
Delegate presentation away from BlogDetailsViewController

Code outside BlogDetailsViewController relies on internal showDetailViewForSubsection method to present different blog details views. However, BlogDetailsViewController is not always the currently presented view controller, thus the calls are not always functional.

- Creating BlogDetailsPresentationDelegate
- Using delegate calls instead of direct presentation calls

staskus committed:
Present blog details subsection view on the selected section

- Implement BlogDetailsPresentationDelegate on MySiteViewController
- When BlogDetails asks to present one of the sections check the currently selected section
- Present on blogDashboardViewController or blogDetailsViewController based on selection

staskus commented:
Added a fix for the navigation issue when "Home" is the initial screen. Additionally discovered that the same issue exists when opening "Stats" from Home Screen using 3DTouch. 

CC @momo-ozawa @guarani 

## Issue

**Rootcause**

The problem is that 3DTouch / Widget routes depend on `showBlogDetails` and `BlogDetailsSubsection`. So navigation logic outside `BlogDetailsViewController` is coupled with the functionality inside `BlogDetailsViewController`. 

**Possible fix**

A possible quick-fix for Stats would be to have some if/switch statements inside `MySiteViewController`:

```swift
  func showBlogDetailsSubsection(_ subsection: BlogDetailsSubsection) {
      guard let blog = blog else {
          return
      }
      
      if currentSection == .dashboard, subsection == .posts, let dashboardVC = blogDashboardViewController {
          StatsViewController.show(for: blog, from: dashboardVC)
      } else {
          blogDetailsViewController?.showDetailView(for: subsection)
      }
  }
```

However, we would need to have if/swiftch statement for each of the `subsection` that we want to present on top of `blogDashboardViewController`. I found this solution to be error-prone and hard to maintain.

**Proposed fix**

Since we want to reliably present `BlogDetails subsections` in a different context, I propose a solution to reflect that. Move away presentation logic from `BlogDetailsViewController` to `MySiteViewController`.

- Create `BlogDetailsPresentationDelegate`
- `MySitesViewController` become `BlogDetailsPresentationDelegate` and present BlogDetail-related controllers either on `Dashboard` or on `BlogDetails` views.

This way we will support all existing routing cases as well as any other additional BlogDetails subsections that we want to present from outside. 

## Testing instructions 

**1st test case**
Test the same scenario from the description with these configurations:
1. iPhone + My Site > App Settings > Initial Screen -> Home
2. iPad + My Site > App Settings > Initial Screen -> Home
3. iPhone + My Site > App Settings > Initial Screen -> Menu
4. iPad +  My Site > App Settings > Initial Screen -> Menu

## Regression Notes

1. Potential unintended areas of impact

Menu section (BlogDetailsViewController)

2. What I did to test those areas of impact

- Open app
- Open "Menu tab"
- Make sure all sections (Stats, Activity, etc..) are opened when selecting manually.

## Videos

**iPhone - Open from Widget - Home**

https://user-images.githubusercontent.com/4062343/174043761-3a5af703-2ce0-4806-962c-708b9fdcd27c.mov

**iPhone - Open from Widget - Menu**

https://user-images.githubusercontent.com/4062343/174044373-9f695663-804b-4a75-a72c-21598681817c.mov

**iPhone - 3D Touch - Home**

https://user-images.githubusercontent.com/4062343/174043981-f1d02a9f-4fb9-409d-b9b6-8487d73e4118.mov

**iPad - Open from Widget - Home**

https://user-images.githubusercontent.com/4062343/174044561-4e1e129e-9811-4010-8c8a-fd7683032f39.mp4

momo-ozawa reviewed (approved):
I tested the following cases and it works as described.
- iPhone + My Site > App Settings > Initial Screen -> Home
- iPad + My Site > App Settings > Initial Screen -> Home
- iPhone + My Site > App Settings > Initial Screen -> Menu
- iPad + My Site > App Settings > Initial Screen -> Menu

> Since we want to reliably present BlogDetails subsections in a different context, I propose a solution to reflect that. Move away presentation logic from BlogDetailsViewController to MySiteViewController.

Great solution! LGTM ✨

guarani commented:
👋 @staskus, I re-triggered CI here via https://github.com/wordpress-mobile/WordPress-iOS/pull/18869.

guarani commented:
I also added a milestone and labels. Please let us know here when this can be merged, or if any CI issues come up, etc. 🙇

staskus commented:
> I also added a milestone and labels. Please let us know here when this can be merged, or if any CI issues come up, etc. 🙇

@guarani thanks! 

All checks have passed. Testing and reviews are done. It's safe to merge.

guarani commented:
>  It's safe to merge.

@staskus, I merged this now 🚀 Thanks!
